### ISSUE TITLE
Many-to-One Sequence Dataset

### ISSUE THREAD

**Issue by pindapuj** [rusty1s/pytorch_geometric#1347]
## ❓ Questions & Help

Hi!  Thank you for the great package!

I just started using Pytorch_geometric, and have a question about a dataloader use-case. 
Specifically, I have time-series of graphs that correspond to a single label. 

So, one sample would be [(g_1, g_2, ...g_t),label_0]. I was wondering the correct way to load such a dataset. This closest issue I found was[ #321](https://github.com/rusty1s/pytorch_geometric/issues/321). However, I'm still quite unsure what you mean by "encode multiple graphs" in one `Data` object, and how to use `seq` assignments. The nodes that occur in one time-stamp have semantic meaning when they occur in the next time stamp, so I have to make sure they are aligned. This is my best guess: 

```
Let each g_i be a networkx graph object:
seq_len=10
series = [nx.Graph(g_1)....nx.Graph(g_10)]
data_objects = [torch_geometric.utils.convert.from_networkx(graph) for graph in series]

#get and combine edgelist,ditto for nodes, and feautres
#this is quite inefficient, probably better way! 
new_edge_list = sum([graph.edge_list for graph in series],[]) #get one long list
new_node_feature_matrix = np.stack(np.asarray([graph.x for graph in series])

#what is sequence assignment vector? 
seq = [[i] * g.number_of_nodes() for i, g in enumerate(series)]

data_1 = Data(x=new_node_feature_matrix,edge_index=new_edge_index,y=[0])
data.seq = seq
```
Now, that we have one sequence converted, I would repeat the process for the other samples. 
To obtain a list of `Data` objects that can then be feed into a  `Dataloader`. But, now I am unsure how to restore the original graph (g_i) for getting say a sub-graph level embedding. 

Any help or direction would be very much appreciated!!

**Comment by rusty1s** [comment]
The `DataLoader` should batch different sequences together, so encoding your sequence in a single example is a good way to go. Your code looks mostly correct, however, edge indices of different timestamps shouldn't overlap.
```
cumsum = 0
new_edge_index = []
for i in range(seq_len):
    new_edge_index.append(data_objects.edge_index + cumsum)
    cumsum += data.num_nodes
new_edge_index = torch.cat(new_edge_index, dim=1)
```
Your `seq` vector does the right job. You can use it as
```python
seq_idx = data.seq + data.batch * seq_len
x = global_pool(x, seq_idx).view(batch_size, seq_len, num_features)
```
inside your model.

**Comment by pindapuj** [comment]
Excellent, thank you so much! 

Quick follow-up: Is it possible to take the gradients with respect to inputs? Conceptually, I'd like to take the gradients wrt to the node features, and the original adjacency matrix. In just torch, this could be done by:
```
temp_x  = torch.autograd.Variable(data.x.clone(),requires_grad=True)

#Error! 
temp_edge = torch.autograd.Variable(data.edge_index.clone(),requires_grad=True)
output = model(x,temp_edge)
#get loss and compute backward pass.
```

However, its unclear what to do for pytorch_geometric. A `Variable` requires that the tensor be `torch.Floating` but `Data` objects (and GCNConv layers etc) expect type to be `torch.LongTensor/torch.long`. I'm wondering if there is a work around. 

Thank you so much!

**Comment by rusty1s** [comment]
You can see my reply in #1361. In general, you can compute gradients with respect to inputs such as node features and edge weights. However, as you correctly mentioned, you cannot compute them for `edge_index`. However, it can be computed by representing the edges as a tuple of `(edge_index, edge_weight)` (and most GNNs allow to take in `edge_weight` as an additional argument). You can then require gradients computation for `edge_weight`.

**Comment by pindapuj** [comment]
Thank you!
